docs_src/design/adr/013-Device-Service-Events-Message-Bus.md

```diff
@@ -0,0 +1,247 @@
+
+Currently Core Data publishes Events to the simple `events` topic. All Application Services running receive every Event published, whether they want them or not. The Events can be filtered out using the `FilterByDeviceName` pipeline function, but the Application Services still receives every Event and process all the Events to some extent. This could cause load issues in a deployment with many devices and large volume of Events from various devices or a very verbose device that the Application Services is not interested in.
+
+> *Note: The current `FilterByDeviceName` is only good if the device name is known statically and the only instance of the device defined by the `DeviceProfileName`. What we really need is `FilterByDeviceProfileName` which allows multiple instances of a device to be filtered for, rather than a single instance as it it now. The V2 API will be adding `DeviceProfileName` to the Events, so in Ireland this  filter will be possible.*
+
+Pub/Sub systems have advanced topic schema, which we can take advantage of from Application Services to filter for just the Events the Application Service actual wants. Publishers of Events must add the `DeviceProfileName` and/or `DeviceName` to the topic in the form `edgex/events/<device-profile-name>/<device-name>`. This allows Application Services to filter for just the Events from the device(s) it wants by only subscribing to those `DeviceProfileNames` or the specific `DeviceNames`. Example subscribe topics if above schema is used:
+
+- **edgex/events/#**
+  - All Events 
+  - Core Data will subscribe using this topic schema
+- **edgex/events/Random-Integer-Device/#** 
+  - Only Events from devices created from the **Random-Integer-Device** device profile
+- **edgex/events/Random-Integer-Device/Random-Integer-Device1** 
+  - Only Events from the **Random-Integer-Device1** Device
+
+The MessageBus abstraction allows for multiple subscriptions, so an Application Service could specify to receive data from multiple specific device profiles or devices by creating multiple subscriptions. i.e.  `edgex/Events/Random-Integer-Device/#` and  `edgex/Events/Random-Boolean-Device/#`. Currently the App SDK only allows for a single subscription topic to be configured, but that could easily be expanded to handle a list of subscriptions. See [Configuration](#configuration) section below for details. 
+
+Core Data's existing publishing of Events would also need to be changed to use this new topic schema. One challenge with this is Core Data doesn't currently know the `DeviceProfileName` or `DeviceName` when it receives a CBOR encoded event. This is because it doesn't decode the Event until after it has published it to the MessageBus. The V2 API will be enhanced to change the AddEvent endpoint from `/event` to `/event/{profile}/{device}` so that `DeviceProfileName` and `DeviceName` are always know no matter how the request is encoded.
```

Comment:
I don't see the necessity of modifying the endpoint route.
If Core Data needs to persist the CBOR event, it has to encode it first.
Also, we have to validate whether the Device Name and Profile Name in the Event body is the same as the URL parameters.

Reply:
No, CBOR data is published before Core Data decodes it. Yes, you are correct, the Device Name and Profile Name will need to be validate that they match.